sql/analyzer/resolve_subqueries.go

```diff
@@ -250,48 +272,86 @@ func isDeterminstic(n sql.Node) bool {
 	return res
 }
 
-// cacheSubqueryResults determines whether it's safe to cache the results for any subquery expressions, and marks the
+// cacheSubqueryResults determines whether it's safe to cache the results for subqueries (expressions and aliases), and marks the
 // subquery as cacheable if so. Caching subquery results is safe in the case that no outer scope columns are referenced,
 // if all expressions in the subquery are deterministic, and if the subquery isn't inside a trigger block.
-func cacheSubqueryResults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+func cacheSubqueryResults(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
 	// No need to inspect for trigger blocks as the Analyzer is recursively invoked on trigger blocks.
-	if n, ok := n.(*plan.TriggerBeginEndBlock); ok {
+	if n, ok := node.(*plan.TriggerBeginEndBlock); ok {
 		return n, transform.SameTree, nil
 	}
-	return transform.NodeExprsWithNode(n, func(n sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
-		s, ok := e.(*plan.Subquery)
-		if !ok || !s.Resolved() {
-			return e, transform.SameTree, nil
-		}
 
-		scopeLen := len(scope.newScope(n).Schema())
-		cacheable := nodeIsCacheable(s.Query, scopeLen)
+	return transform.NodeWithCtx(node, nil, func(context transform.Context) (sql.Node, transform.TreeIdentity, error) {
+		if sqa, ok := context.Node.(*plan.SubqueryAlias); ok {
+			subScope := newScopeWithDepth(scope.RecursionDepth() + 1)
+			if scope != nil && len(scope.nodes) > 0 {
+				if scope.CurrentNodeIsSubqueryExpression {
+					subScope.nodes = append(subScope.nodes, scope.InnerToOuter()...)
+				}
+			}
+			if nodeIsCacheable(ctx, sqa.Child, subScope) {
+				return sqa.WithCachedResults(), transform.NewTree, nil
+			}
+			return context.Node, transform.SameTree, nil
+		} else if expressioner, ok := context.Node.(sql.Expressioner); ok {
+			exprs := expressioner.Expressions()
```

Comment:
Same comment about using transform.NodeExprs (or OneNodeExprsWithNode if you don't want it recursively applied)